lib/writable.js

```diff
@@ -1,69 +1,227 @@
-  // Returns: boolean
-  //   true - can write more
-  //   false - buffer overflow, do not write more
-) {
-  if (typeof(chunk) === 'string') chunk = Buffer.from(chunk, encoding);
-  if (callback) callback();
-};
-
-Writable.prototype.end = function(chunk, encoding, callback) {
-  this.write(chunk, encoding, callback);
-};
-
-Writable.prototype.destroy = function(error) {
-  throw error;
-};
-
-Writable.prototype.cork = function() {
-  this.corked++;
-};
-
-Writable.prototype.uncork = function() {
-  if (this.corked) this.corked--;
-};
-
-
-Writable.prototype.setDefaultEncoding = function(encoding) {
-  this.encoding = encoding;
-};
+const writeChunkFuncName = Symbol('writeChunk');
+const errorFuncName = Symbol('error');
+
+class WritableBuffer {
+  constructor(highWaterMark) {
+    this.offset = 0;
+    this.queue = [];
+    this.writableLength = 0;
+    this.highWaterMark = highWaterMark;
+    this.buffer = Buffer.alloc(highWaterMark);
+  }
+
+  [writeChunkFuncName](chunk) {
+    chunk.copy(this.buffer, this.offset);
+    this.writableLength += chunk.length;
+    this.offset += chunk.length;
+  }
+
+  write(chunk, encoding) {
+    const remainingSize = this.highWaterMark - this.offset;
+
+    if (chunk.length <= remainingSize) {
+      this[writeChunkFuncName](chunk, encoding);
+    } else {
+      const currentBufferChunk = chunk.slice(0, this.remainingSize);
+      this[writeChunkFuncName](currentBufferChunk);
+      this.queue.push(this.buffer);
+      this.buffer = Buffer.alloc(this.highWaterMark);
+      this.offset = 0;
+
+      const nextBufferChunk = chunk.slice(remainingSize);
+
+      if (nextBufferChunk.length < this.buffer.length) {
+        this[writeChunkFuncName](nextBufferChunk, encoding);
+      } else {
+        this.queue.push(nextBufferChunk);
+        this.writableLength += nextBufferChunk.length;
+      }
+    }
+  }
+
+  getBuffer() {
+    const bufferData = this.buffer.slice(0, this.offset);
+    this.queue.push(bufferData);
+    const ret = this.queue;
+    this.queue = [];
+    this.offset = 0;
+    this.writableLength = 0;
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    defaultEncoding = 'utf-8',
+    highWaterMark = DEFAULT_HIGH_WATERMARK
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+
+    if (!this._write && !this._writev) {
+      throw new Error('[WRITE_AND_WRITEV_NOT_IMPLEMENTED]');
+    }
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.highWaterMark = highWaterMark;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(highWaterMark);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (!this.corked && this.writableLength) {
```

Comment:
```suggestion
    if (this.corked === 0 && this.writableLength) {
```
Also, you should also add check to avoud this.corked become <0.